web-common-marlowe/src/Decode/helpers.purs

```diff
@@ -0,0 +1,24 @@
+module Decode.Helpers where
+
+import Prelude
+import Control.Monad.Except (ExceptT(..))
+import Data.Either (Either(..))
+import Data.Identity (Identity(..))
+import Foreign (F)
+
+-- This is a lazy version of the Alt operator to be used in the context of decoding a Foreing
+-- value into a PureScript data type.
+-- The problem of using <|> from Control.Alt is that `F a` is defined as an `Except MultipleErrors a`
+-- which resolves to `ExceptT MultipleErrors Indentity a`, and the Identity monad is not lazy, so when
+-- we do:
+-- parseAsX a <|> parseAsY a
+-- the value will be parsed eagerly as X and Y before discarding one option.
+-- With this helper, the parsers needs to be defined as:
+-- parseAsX a <|> \_ -> parseAsY a
+-- but then the parseAsY will only be invoked if the parseAsX failed.
+altLazy :: forall a. F a -> (Unit -> F a) -> F a
+altLazy (ExceptT (Identity (Right a))) _ = pure a
+
+altLazy _ b = b unit
```

Comment:
That is beautifully satisfying. But why don't we do:

```
altLazy :: forall a. F a -> F a -> F a
altLazy a b = altLazy' a (\_ -> b)
  where
  altLazy' :: forall a. F a -> (Unit -> F a) -> F a
  [etc.]
```

?

Then we wouldn't need to change anything else. Also I think I'd prefer to put this in `web-common`, as `Control.Alt.Extra`. Or possibly `Control.Alt.Lazy`... ? :thinking:

Reply:
Hm, it would be cool, but I am not sure that is going to work, because I am guessing it is having the closure where it is that prevents the evaluation. Otherwise it would still get reduced before calling Alt. But it is easy to test, worth a try

Reply:
That's a good point, I think you might be right. Oh well.

Reply:
I tried this and it didn't work, you need to have the explicit `\_ ->` in the parser definition to make it lazy. Or eventually we could try again to use Control.Lazy or Data.Lazy, but for the moment I think this is the simplest solution